patches/api/0323-Add-BlockDropResourcesEvent-and-subclasses.patch

```diff
@@ -0,0 +1,196 @@
++        return source;
++    }
++}
+diff --git a/src/main/java/io/papermc/paper/event/block/BlockDropResourcesEvent.java b/src/main/java/io/papermc/paper/event/block/BlockDropResourcesEvent.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..682d6fac5e65b6c45aaa230bd204762308b516e5
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/event/block/BlockDropResourcesEvent.java
+@@ -0,0 +1,93 @@
++package io.papermc.paper.event.block;
++
++import com.google.common.base.Preconditions;
++import org.bukkit.block.Block;
++import org.bukkit.block.BlockState;
++import org.bukkit.event.HandlerList;
++import org.bukkit.event.block.BlockEvent;
++import org.bukkit.inventory.ItemStack;
++import org.jetbrains.annotations.NotNull;
++import org.jetbrains.annotations.Nullable;
++
++import java.util.List;
++
++/**
++ * Called when a block drops items in a world as a result of being broken.
++ * The block will exist in the world at the time of firing.
++ * <p>
++ * This event fires in between {@link org.bukkit.event.block.BlockBreakEvent BlockBreakEvent}
++ * and {@link org.bukkit.event.block.BlockDropItemEvent BlockDropItemEvent}.
++ * </p>
++ * In rare circumstances, the {@link #getBlock() block} may be removed from the world before this
++ * event is called.
++ * For this reason, it is recommended to use {@link #getBlockState()} to fetch information about the
++ * broken block.
++ */
++public class BlockDropResourcesEvent extends BlockEvent {
++
++    private static final HandlerList HANDLER_LIST = new HandlerList();
++
++    private final List<ItemStack> drops;
++    private final BlockState blockState;
++
++    /**
++     * Creates a BlockDropResourcesEvent from a block that has not been removed from the world.
++     * <p>
++     * If the block might already be removed from the world then the
++     * {@link #BlockDropResourcesEvent(Block, BlockState, List) full constructor}
++     * should be used instead.
++     *
++     * @param block the block dropping resources
++     * @param drops the resources being dropped
++     */
++    public BlockDropResourcesEvent(@NotNull Block block, @NotNull List<ItemStack> drops) {
++        this(block, null, drops);
++    }
++
++    /**
++     * Creates a BlockDropResourcesEvent from a block and, optionally, a snapshot of the block
++     * created before it was (or wasn't) removed from the world.
++     * If a snapshot is not provided then one will be created from {@link Block#getState()}.
++     *
++     * @param block      the block dropping resources
++     * @param blockState a snapshot of the block created before removal from the world
++     * @param drops      the resources being dropped
++     */
++    public BlockDropResourcesEvent(@NotNull Block block, @Nullable BlockState blockState, @NotNull List<ItemStack> drops) {
++        super(block);
++        this.blockState = blockState != null ? blockState : block.getState(true);
++        this.drops = Preconditions.checkNotNull(drops, "drops");
```

Comment:
since you guarantee mutability it'd be good to just do new ArrayList() around the passed in argument.

Reply:
good catch; fixed